File: Contacts/PHP/Cli/contacts.php
<?php

function validateContact(string $name, string $phone, string $email): array {
    $errors = [];
    $name = trim($name);
    $phone = trim($phone);
    $email = trim($email);
    if ($name === '') {
        $errors['name'] = 'Name is required';
    } elseif (strlen($name) < 2 || strlen($name) > 100 || !preg_match('/^[A-Za-zÀ-ÿ\' .-]+$/', $name)) {
        $errors['name'] = 'Name must be 2-100 characters (letters, spaces, apostrophes, hyphens, dots)';
    }
    if ($phone === '') {
        $errors['phone'] = 'Phone is required';
    } elseif (!preg_match('/^[0-9 +().-]{7,20}$/', $phone)) {
        $errors['phone'] = 'Phone must be 7-20 characters (digits, spaces, +, parentheses, dashes)';
    }
    if ($email === '') {
        $errors['email'] = 'Email is required';
    } elseif (!preg_match('/^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/', $email)) {
        $errors['email'] = 'Invalid email format';
    }
    return $errors;
}

function addContact(array &$contacts, string $name, string $phone, string $email): void {
    $errors = validateContact($name, $phone, $email);
    if ($errors) {
        foreach ($errors as $error) {
            fwrite(STDERR, $error . "\n");
        }
        return;
    }
    $contacts[] = ["name" => trim($name), "phone" => trim($phone), "email" => trim($email)];
    echo "Contact added!\n";
}

function listContacts(array $contacts): void {
    if (empty($contacts)) {
        echo "No contacts found.\n";
        return;
    }
    foreach ($contacts as $i => $c) {
        echo "$i. {$c['name']} | {$c['phone']} | {$c['email']}\n";
    }
}

function searchContacts(array $contacts, string $query): array {
    $results = [];
    foreach ($contacts as $c) {
        if (stripos($c['name'], $query) !== false) {
            $results[] = $c;
        }
    }
    if (empty($results)) {
        echo "No contacts found.\n";
    } else {
        foreach ($results as $i => $c) {
            echo "$i. {$c['name']} | {$c['phone']} | {$c['email']}\n";
        }
    }
    return $results;
}

function deleteContact(array &$contacts, int $index): void {
    if (isset($contacts[$index])) {
        $removed = $contacts[$index];
        array_splice($contacts, $index, 1);
        echo "Deleted {$removed['name']}\n";
    } else {
        echo "Invalid index.\n";
    }
}

function main(): void {
    $contacts = [];
    while (true) {
        echo "\n--- Contact Manager ---\n";
        echo "1. Add Contact\n";
        echo "2. List Contacts\n";
        echo "3. Search Contacts\n";
        echo "4. Delete Contact\n";
        echo "5. Exit\n";
        echo "Choose an option: ";
        $choice = trim(fgets(STDIN));
        switch ($choice) {
            case '1':
                echo "Name: ";
                $name = trim(fgets(STDIN));
                echo "Phone: ";
                $phone = trim(fgets(STDIN));
                echo "Email: ";
                $email = trim(fgets(STDIN));
                addContact($contacts, $name, $phone, $email);
                break;
            case '2':
                listContacts($contacts);
                break;
            case '3':
                echo "Search query: ";
                $query = trim(fgets(STDIN));
                searchContacts($contacts, $query);
                break;
            case '4':
                listContacts($contacts);
                echo "Enter index to delete: ";
                $idx = (int) trim(fgets(STDIN));
                deleteContact($contacts, $idx);
                break;
            case '5':
                echo "Goodbye!\n";
                exit(0);
        }
    }
}

if (PHP_SAPI === 'cli' && isset($argv[0]) && realpath($argv[0]) === __FILE__) {
    main();
}

File: Contacts/Elixir/Cli/lib/contacts.ex
defmodule Contacts do
  @moduledoc """
  Contact management core: validation, add, list, search, update and delete
  in-memory contacts.

  Mirrors the Ruby and Node CLI behaviour:

  - name: required, 2..100 chars, letters/spaces/apostrophes/hyphens/dots
  - phone: required, 7..20 chars, digits/spaces/+/(/)/dashes
  - email: required, matches `user@domain.tld`

  Return style:
    {:ok, contact}   on success
    {:error, msg}    on failure (msg string)
  """

  @name_re ~r/\A[A-Za-zÀ-ÿ' .-]+\z/
  @phone_re ~r/\A[0-9 +().-]{7,20}\z/
  @email_re ~r/\A[^\s@]+@[^\s@]+\.[^\s@]{2,}\z/

  defstruct [:id, :name, :phone, :email]

  @type t :: %__MODULE__{id: pos_integer, name: String.t(), phone: String.t(), email: String.t()}

  @spec validate(name :: String.t(), phone :: String.t(), email :: String.t()) :: [String.t()]
  def validate(name, phone, email) do
    []
    |> maybe_name_error(name)
    |> maybe_phone_error(phone)
    |> maybe_email_error(email)
  end

  defp maybe_name_error(errors, name) do
    cond do
      name == "" -> ["Name is required" | errors]
      String.length(name) < 2 or String.length(name) > 100 ->
        ["Name must be 2-100 characters (letters, spaces, apostrophes, hyphens, dots)" | errors]
      not Regex.match?(@name_re, name) ->
        ["Name must be 2-100 characters (letters, spaces, apostrophes, hyphens, dots)" | errors]
      true -> errors
    end
  end

  defp maybe_phone_error(errors, phone) do
    cond do
      phone == "" -> ["Phone is required" | errors]
      not Regex.match?(@phone_re, phone) ->
        ["Phone must be 7-20 characters (digits, spaces, +, parentheses, dashes)" | errors]
      true -> errors
    end
  end

  defp maybe_email_error(errors, email) do
    cond do
      email == "" -> ["Email is required" | errors]
      not Regex.match?(@email_re, email) -> ["Invalid email format" | errors]
      true -> errors
    end
  end

  @spec create([t()], String.t(), String.t(), String.t()) ::
          {:ok, t()} | {:error, [String.t()]}
  def create(contacts, name, phone, email) do
    case validate(name, phone, email) do
      [] ->
        id = (Enum.map(contacts, & &1.id) |> Enum.max(fn -> 0 end)) + 1
        {:ok, %__MODULE__{id: id, name: name, phone: phone, email: email}}

      errors ->
        {:error, errors}
    end
  end

  @spec list([t()]) :: [t()]
  def list(contacts), do: contacts

  @spec search([t()], String.t()) :: [t()]
  def search(contacts, query) do
    q = String.downcase(query)
    Enum.filter(contacts, &String.contains?(String.downcase(&1.name), q))
  end

  @spec update([t()], pos_integer, String.t(), String.t(), String.t()) ::
          {:ok, t()} | {:error, [String.t()]} | {:error, :not_found}
  def update(contacts, id, name, phone, email) do
    if Enum.any?(contacts, &(&1.id == id)) do
      case validate(name, phone, email) do
        [] -> {:ok, %__MODULE__{id: id, name: name, phone: phone, email: email}}
        errors -> {:error, errors}
      end
    else
      {:error, :not_found}
    end
  end

  @spec delete([t()], pos_integer) :: {:ok, t()} | {:error, :not_found}
  def delete(contacts, id) do
    case Enum.find(contacts, &(&1.id == id)) do
      nil -> {:error, :not_found}
      contact -> {:ok, contact}
    end
  end
end